File: src/main/java/com/github/mgramin/sqlboot/model/resource_type/impl/composite/md/CustomVisitor.java
/*
 * The MIT License (MIT)
 * <p>
 * Copyright (c) 2016-2019 Maksim Gramin
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.github.mgramin.sqlboot.model.resource_type.impl.composite.md;

import org.commonmark.node.AbstractVisitor;
import org.commonmark.node.FencedCodeBlock;
import org.commonmark.node.Heading;
import org.commonmark.node.Text;

import java.util.LinkedHashMap;
import java.util.Map;

public class CustomVisitor extends AbstractVisitor {

    private String currentTag;
    private Map<String, String> map = new LinkedHashMap<>();

    @Override
    public void visit(Text text) {
        if (text.getParent() instanceof Heading && ((Heading)text.getParent()).getLevel()>=3) {
            currentTag = text.getLiteral();
        }
    }

    @Override
    public void visit(FencedCodeBlock fencedCodeBlock) {
        if (fencedCodeBlock.getFenceLength() == 4) {
            map.put(currentTag, fencedCodeBlock.getLiteral());
        }
    }

    public Map<String, String> getMap() {
        return map;
    }

}

File: src/main/java/com/github/mgramin/sqlboot/model/resource_type/impl/composite/md/CustomVisitor.kt
/*
 * The MIT License (MIT)
 * <p>
 * Copyright (c) 2016-2019 Maksim Gramin
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.github.mgramin.sqlboot.model.resource_type.impl.composite.md

import org.commonmark.node.AbstractVisitor
import org.commonmark.node.FencedCodeBlock
import org.commonmark.node.Heading
import org.commonmark.node.Text

import java.util.LinkedHashMap

class CustomVisitor : AbstractVisitor() {

    private var currentTag: String = ""
    val map = LinkedHashMap<String, String>()

    override fun visit(text: Text) {
        if (text.parent is Heading && (text.parent as Heading).level >= 3) {
            currentTag = text.literal
        }
    }

    override fun visit(fencedCodeBlock: FencedCodeBlock) {
        if (fencedCodeBlock.fenceLength == 4) {
            map[currentTag] = fencedCodeBlock.literal
        }
    }

    fun getMap(): Map<String, String> {
        return map
    }

}
